hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/utils/MockKeyedStateStore.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.state.MapState;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.api.common.state.ReducingState;
+import org.apache.flink.api.common.state.ReducingStateDescriptor;
+import org.apache.flink.api.common.state.ValueState;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.hudi.sink.utils.keyedstate.MockKeyContext;
+import org.apache.hudi.sink.utils.keyedstate.MockKeyedListState;
+import org.apache.hudi.sink.utils.keyedstate.MockKeyedMapState;
+import org.apache.hudi.sink.utils.keyedstate.MockKeyedValueState;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+/**
+ * An {@link KeyedStateStore} for testing purpose.
+ */
+@SuppressWarnings("rawtypes")
+public class MockKeyedStateStore implements KeyedStateStore {
+  private final MockKeyContext keyContext;
+
+  private Map<String, MockKeyedValueState> keyedValueStateMap;
+  private Map<String, MockKeyedListState> keyedListStateMap;
+  private Map<String, MockKeyedMapState> keyedMapStateMap;
+
+  private final Map<Long, Map<String, MockKeyedValueState>> historyValueStateMap;
+  private final Map<Long, Map<String, MockKeyedListState>> historyListStateMap;
+  private final Map<Long, Map<String, MockKeyedMapState>> historyMapStateMap;
+
+  private Map<String, MockKeyedValueState> lastSuccessValueStateMap;
+  private Map<String, MockKeyedListState> lastSuccessListStateMap;
+  private Map<String, MockKeyedMapState> lastSuccessMapStateMap;
+
+  public MockKeyedStateStore() {
+    keyContext = new MockKeyContext();
+
+    this.keyedValueStateMap = new HashMap<>();
+    this.keyedListStateMap = new HashMap<>();
+    this.keyedMapStateMap = new HashMap<>();
+
+    this.historyValueStateMap = new HashMap<>();
+    this.historyListStateMap = new HashMap<>();
+    this.historyMapStateMap = new HashMap<>();
+
+    this.lastSuccessValueStateMap = new HashMap<>();
+    this.lastSuccessListStateMap = new HashMap<>();
+    this.lastSuccessMapStateMap = new HashMap<>();
+  }
+
+  @Override
+  public <T> ValueState<T> getState(ValueStateDescriptor<T> stateProperties) {
+    String name = stateProperties.getName();
+    keyedValueStateMap.putIfAbsent(name, new MockKeyedValueState<T>(keyContext));
+    return keyedValueStateMap.get(name);
+  }
+
+  @Override
+  public <T> ListState<T> getListState(ListStateDescriptor<T> stateProperties) {
+    String name = stateProperties.getName();
+    keyedListStateMap.putIfAbsent(name, new MockKeyedListState<T>(keyContext));
+    return keyedListStateMap.get(name);
+  }
+
+  @Override
+  public <T> ReducingState<T> getReducingState(ReducingStateDescriptor<T> stateProperties) {
+    throw new UnsupportedOperationException();
```

Comment:
Let's add some error msg like "getReducingState is not supported yet"

Reply:
Thank you for helping me improve the code standards